string.c

```diff
@@ -3326,27 +3326,33 @@ rb_str_append(VALUE str, VALUE str2)
     return rb_str_buf_append(str, str2);
 }
 
+#if !USE_RVARGC
 #define MIN_PRE_ALLOC_SIZE 48
+#endif
 
 MJIT_FUNC_EXPORTED VALUE
 rb_str_concat_literals(size_t num, const VALUE *strary)
 {
     VALUE str;
-    size_t i, s;
-    long len = 1;
+    size_t i, s = 0;
+    unsigned long len = 1;
 
     if (UNLIKELY(!num)) return rb_str_new(0, 0);
     if (UNLIKELY(num == 1)) return rb_str_resurrect(strary[0]);
 
     for (i = 0; i < num; ++i) { len += RSTRING_LEN(strary[i]); }
+
+#if USE_RVARGC
+    if (len <= rb_gc_obj_slot_size(strary[0]) - sizeof(struct RBasic)) {
+#else
     if (LIKELY(len < MIN_PRE_ALLOC_SIZE)) {
+#endif
         str = rb_str_resurrect(strary[0]);
         s = 1;
     }
     else {
         str = rb_str_buf_new(len);
         rb_enc_copy(str, strary[0]);
-        s = 0;
     }
```

Comment:
With this change, I get similar performance for your "Small string interpolation" benchmark and much better performance on the benchmark shown below.

```suggestion
    str = rb_str_buf_new(len);
    str_enc_copy_direct(str, strary[0]);
```

Benchmark:

```ruby
a = "Hello"
b = "a" * 30

t.times do
  "#{a}, #{b}!"
end
```

On master:

```
                          6.978M (± 2.1%) i/s -     35.172M in   5.042775s
```

Before the suggested change:

```
                          8.430M (± 1.9%) i/s -     42.651M in   5.061307s
```

After the suggested change:

```
                          9.265M (± 1.8%) i/s -     46.510M in   5.021798s
```